# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [5]:
df[:4]

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [7]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [12]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [17]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [24]:
df["gross income"].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

 ### Аналіз 75-го персентиля

75% клієнтів мають прибуток менше 22.45. І тільки 25% мають прибуток більше за це значення. 
Тобто 3/4 клієнтів приносять сталий, але невеликий дохід. І лише 25% клієнтів приносять найбільший дохід. 
Бізнес залежить від меншої кількості клієнтів, які приносять більший прибуток.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

Очікуваний результат:

In [31]:
df.select_dtypes('object').columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Date', 'Time', 'Payment'],
      dtype='object')

In [33]:
df.City.unique()

array(['Yangon', 'Naypyitaw', 'Mandalay'], dtype=object)

In [35]:
df.City.value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [39]:
df.City.value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекомендації на основі своїх досліджень.

In [45]:
filter_city = (df.City == 'Yangon') | (df.City == 'Mandalay')

In [47]:
df[df.City =='Yangon']["gross income"].mean()

14.874001470588235

In [49]:
df[df.City =='Mandalay']["gross income"].mean()

15.232024096385542

In [51]:
mean_gross_income_filtered = (df[df.City =='Yangon']["gross income"].mean(), df[df.City =='Mandalay']["gross income"].mean())

In [53]:
mean_gross_income_filtered

(14.874001470588235, 15.232024096385542)

In [84]:
mean_gross_income_filtered_text = (
    f"Mean gross income in Yangon = {mean_gross_income_filtered[0]:.3f}\n"
    f"Mean gross income in Mandalay = {mean_gross_income_filtered[1]:.3f}"
)

In [86]:
print(mean_gross_income_filtered_text)

Mean gross income in Yangon = 14.874
Mean gross income in Mandalay = 15.232


![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

## 7.2 Гіпотези стосовно чому в одному місті gross income вище

### Гіпотеза 1. 
У Mandalay люди купують дорожчі товари (вища середня ціна за одиницю).

In [99]:
df[df['City'] == 'Mandalay']['Unit price'].mean()

55.65927710843374

In [101]:
df[df['City'] == 'Yangon']['Unit price'].mean()

54.78085294117646

Висновок: 
Гіпотеза підтвердилась. У Mandalay купують дещо дорожчі товари, що може частково пояснювати вищий gross income.

Середня ціна за одиницю товару (Unit price) у містах Mandalay та Yangon:
- Mandalay: 55.66
- Yangon: 54.78


Рекомендації:
- Розширити асортимент товарів преміум-класу в Mandalay.
- Запустити рекламу дорогих товарів.
- Спробувати просунути схожі товари в Yangon, щоб збільшити середній чек.

### Гіпотеза 2.
У Mandalay більше покупок із високою кількістю одиниць товару (Quantity).

In [107]:
df.groupby('City')['Quantity'].mean()

City
Mandalay     5.481928
Naypyitaw    5.582317
Yangon       5.467647
Name: Quantity, dtype: float64

In [111]:
df[df['Quantity'] >= 7].groupby('City').size()
# будемо вважати великі покупки це речі в кількості від 7ми речей

City
Mandalay     128
Naypyitaw    140
Yangon       130
dtype: int64

Висновок: Гіпотеза не підтвердилась. У Yangon навіть трохи більше великих покупок. Це означає, що більший gross income у Mandalay пояснюється іншими факторами.

Рекомендації: 
- Варто перевірити інші чинники: ціни, категорії товарів, або тип клієнтів.
- У Yangon потенціал для зростання прибутку є, можна спробувати підвищити середню ціну за одиницю (через акції на преміум-товари).

### Гіпотеза 3.
У Mandalay більше покупців із типом Member, і вони приносять більше прибутку.

In [119]:
df[df['City'] == 'Mandalay']['Customer type'].value_counts()

Customer type
Normal    167
Member    165
Name: count, dtype: int64

In [121]:
df[df['City'] == 'Yangon']['Customer type'].value_counts()

Customer type
Normal    173
Member    167
Name: count, dtype: int64

In [123]:
df.groupby('Customer type')['gross income'].mean()

Customer type
Member    15.609110
Normal    15.148707
Name: gross income, dtype: float64

Висновок:
Кількість клієнтів за типом у Mandalay та Yangon:
   - Mandalay: 167 Normal, 165 Member
   - Yangon: 173 Normal, 167 Member

Середній gross income за типом клієнта:
   - Member: 15.61
   - Normal: 15.15

Гіпотеза частково підтвердилася. Дійсно, клієнти типу Member приносять більше прибутку. Однак у Mandalay їх не більше, ніж у Yangon, тож саме це не є основним фактором вищого доходу у місті.

Рекомендації:
- Розвивати програму лояльності: стимулювати збільшення частки Member-клієнтів у всіх містах.
- Сегментувати клієнтів і розробити окремі стратегії маркетингу для кожного типу.
- Аналізувати поведінку Member-клієнтів — які категорії товарів вони купують частіше, і просувати ці товари іншим.

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [62]:
df['gross margin percentage'].value_counts()

gross margin percentage
4.761905    1000
Name: count, dtype: int64

In [64]:
if df['gross margin percentage'].nunique() == 1:
    print ("У всіх всіх інвойсах однакова маржа")
else:
    print("Маржа відрізняється для різних інвойсів")

У всіх всіх інвойсах однакова маржа


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Враховуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розподіл кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [68]:
df.query('City == "Yangon"')['Product line'].value_counts(normalize=True)

Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: proportion, dtype: float64

In [70]:
df.query('City == "Mandalay"')['Product line'].value_counts(normalize=True)

Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: proportion, dtype: float64

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)